File: registry-compiler/src/main/java/com/github/fengdai/registry/compiler/BindingSet.java
package com.github.fengdai.registry.compiler;

import com.google.common.collect.ImmutableList;
import com.squareup.javapoet.ClassName;
import java.util.LinkedHashMap;
import java.util.Map;
import javax.lang.model.element.TypeElement;

final class BindingSet {
  final ClassName dataClassName;
  final ImmutableList<Binding> bindings;
  final boolean isMultiBinding;

  private BindingSet(TypeElement dataElement, ImmutableList<Binding> bindings) {
    this.dataClassName = ClassName.get(dataElement);
    this.bindings = bindings;
    this.isMultiBinding = bindings.size() > 1;
  }

  static class Builder {
    private final TypeElement dataElement;
    private final Map<TypeElement, Binding> bindings = new LinkedHashMap<>();

    Builder(TypeElement dataElement) {
      this.dataElement = dataElement;
    }

    void add(Binding binding) {
      if (!binding.dataElement.equals(dataElement)) {
        throw new IllegalStateException();
      }

      final TypeElement viewHolderElement = binding.viewHolderElement;
      final Binding existedBinding = bindings.get(viewHolderElement);
      if (existedBinding != null) {
        throw new DuplicateBindingException(
            String.format("Duplicate binding of %s to %s.\n%s conflicts with %s.",
                binding.dataElement, viewHolderElement, binding.element, existedBinding.element));
      }

      bindings.put(viewHolderElement, binding);
    }

    BindingSet build() {
      return new BindingSet(dataElement, ImmutableList.copyOf(bindings.values()));
    }
  }
}

File: registry-processor/src/main/java/com/github/fengdai/registry/processor/BindingSet.kt
package com.github.fengdai.registry.processor

import com.squareup.javapoet.ClassName
import com.squareup.javapoet.TypeName

data class BindingSet(
  val dataRawType: ClassName,
  val bindings: List<Binding>
) {

  class Builder(private val dataRawType: ClassName) {
    private val bindings = mutableMapOf<TypeName, Binding>()

    fun add(binding: Binding) {
      require(binding.dataRawType == dataRawType)

      val viewHolderType = binding.viewHolderType
      val existedBinding = bindings[viewHolderType]
      if (existedBinding != null) {
        throw DuplicateBindingException("""
          Duplicate binding of ${binding.dataRawType} to $viewHolderType.
          ${binding.targetType} conflicts with
          ${existedBinding.targetType}
          """.trimIndent()
        )
      }
      bindings[viewHolderType] = binding
    }

    fun build(): BindingSet {
      return BindingSet(dataRawType, bindings.values.toList())
    }
  }
}
